pkg/expression/column.go

```diff
@@ -388,11 +388,17 @@ func (col *Column) VecEvalJSON(ctx EvalContext, input *chunk.Chunk, result *chun
 
 const columnPrefix = "Column#"
 
+// StringWithCtx implements Expression interface.
+func (col *Column) StringWithCtx(ctx EvalContext) string {
+	return col.String()
+}
+
 // String implements Stringer interface.
 func (col *Column) String() string {
 	if col.IsHidden && col.VirtualExpr != nil {
 		// A hidden column without virtual expression indicates it's a stored type.
-		return col.VirtualExpr.String()
+		// a virtual column should be ablt to be stringified without context.
+		return col.VirtualExpr.StringWithCtx(nil)
```

Comment:
Could we forbid passing a `nil` as a context if possible? It requires all implements to handle the nil value and the coder may not be aware of it.

Could we introduce a standalone interface like:

```
type ParamValues interface {
    GetParamValue(idx int) types.Datum
}
```

And `EvalContext` can inherit it:

```
type  EvalContext interface {
    ParamValues
}
```

We can also have static implement of it:

```
var EmptyParamValues ParamValues
```

If we know one expression does not need any params, we can call it like this:

```
expr.StringWithCtx(EmptyParamValues)
```

Reply:
Good! Let me try :+1:

Reply:
I realized a problem: the implementation of `StringWithCtx` indeed needs to know whether the argument is `nil` (or any empty value). For `Constant`, the expected behavior is that: if the context is empty, return "?"; if the context is not empty, return the corresponding parameter. Then I cannot get an appropriate implementation of `EmptyParamValues`.

I thought two possible implementations:
1. Return the `"?"` datum. It's not good, because the `"?"` value should be a part of the implementation of `Constant`, but not part of `EmptyParamValues`. Also, I don't want to confuse the empty result and a normal `"?"` string datum.
2. Return `nil` for `*types.Datum`. However,  I don't see the difference between "handling the case when `*types.Datum` is `nil`" and "handling the case when `EvalContext` is `nil`".

Therefore, currently, I prefer to keep the current implementation :thinking: .

Reply:
Could we add new method `ParamValueCount` to `ParamValues` to solve this ?